test/unittest/memory/unittest_memory_pool.cpp

```diff
@@ -0,0 +1,544 @@
+  EXPECT_NO_THROW(pool.planLayout(nntrainer::BasicPlanner()));
+}
+
+/**
+ * @brief plan layout
+ */
+TEST(MemoryPool, plan_layout_03_p) {
+  nntrainer::MemoryPool pool;
+
+  pool.requestMemory(1, 4, 5);
+  EXPECT_NO_THROW(pool.planLayout(nntrainer::BasicPlanner()));
+  EXPECT_EQ(1, pool.size());
+
+  pool.requestMemory(1, 4, 5);
+  EXPECT_NO_THROW(pool.planLayout(nntrainer::BasicPlanner()));
+  EXPECT_EQ(2, pool.size());
+}
+
+/**
+ * @brief deallocate
+ */
+TEST(MemoryPool, deallocate_01_p) {
+  nntrainer::MemoryPool pool;
+
+  EXPECT_NO_THROW(pool.deallocate());
+}
+
+/**
+ * @brief allocate
+ */
+TEST(MemoryPool, allocate_01_n) {
+  nntrainer::MemoryPool pool;
+
+  EXPECT_THROW(pool.allocate(), std::runtime_error);
+
+  pool.requestMemory(1, 4, 5);
+  EXPECT_NO_THROW(pool.planLayout(nntrainer::BasicPlanner()));
+  EXPECT_EQ(1, pool.size());
+
+  EXPECT_NO_THROW(pool.allocate());
+
+  EXPECT_NO_THROW(pool.deallocate());
+}
+
+/**
+ * @brief allocate
+ */
+TEST(MemoryPool, allocate_02_p) {
+  nntrainer::MemoryPool pool;
+
+  pool.requestMemory(1, 4, 5);
+  EXPECT_NO_THROW(pool.planLayout(nntrainer::BasicPlanner()));
+  EXPECT_EQ(1, pool.size());
+
+  EXPECT_NO_THROW(pool.allocate());
+
+  EXPECT_NO_THROW(pool.deallocate());
+}
+
+/**
+ * @brief allocate
+ */
+TEST(MemoryPool, allocate_03_n) {
+  nntrainer::MemoryPool pool;
+
+  pool.requestMemory(1, 4, 5);
+  EXPECT_NO_THROW(pool.planLayout(nntrainer::BasicPlanner()));
+  EXPECT_EQ(1, pool.size());
+
+  EXPECT_NO_THROW(pool.allocate());
+
+  EXPECT_THROW(pool.allocate(), std::runtime_error);
+
+  EXPECT_NO_THROW(pool.deallocate());
+}
+
+/**
+ * @brief allocate
+ */
+TEST(MemoryPool, allocate_04_n) {
+  nntrainer::MemoryPool pool;
+
+  pool.requestMemory(3, 4, 5);
+  EXPECT_NO_THROW(pool.planLayout(nntrainer::BasicPlanner()));
+  EXPECT_EQ(3, pool.size());
+
+  EXPECT_NO_THROW(pool.allocate());
+
+  EXPECT_NO_THROW(pool.deallocate());
+
+  EXPECT_NO_THROW(pool.allocate());
+
+  EXPECT_NO_THROW(pool.deallocate());
+}
+
+/**
+ * @brief size of the pool
+ */
+TEST(MemoryPool, size_01_p) {
+  nntrainer::MemoryPool pool;
+
+  EXPECT_EQ(pool.size(), 0);
+}
+
+/**
+ * @brief size of the pool
+ */
+TEST(MemoryPool, size_02_p) {
+  nntrainer::MemoryPool pool;
+
+  pool.requestMemory(1, 4, 5);
+  EXPECT_EQ(pool.size(), 0);
+}
+
+/**
+ * @brief size of the pool
+ */
+TEST(MemoryPool, size_03_p) {
+  nntrainer::MemoryPool pool;
+
+  pool.requestMemory(1, 4, 5);
+  EXPECT_EQ(pool.size(), 0);
+
+  pool.planLayout(nntrainer::BasicPlanner());
+  EXPECT_EQ(pool.size(), 1);
+
+  pool.allocate();
+  EXPECT_EQ(pool.size(), 1);
+
+  EXPECT_NO_THROW(pool.deallocate());
+}
+
+/**
+ * @brief min requirement
+ */
+TEST(MemoryPool, min_mem_req_01_p) {
+  nntrainer::MemoryPool pool;
+
+  EXPECT_EQ(pool.minMemoryRequirement(), 0);
+}
+
+/**
+ * @brief min requirement
+ */
+TEST(MemoryPool, min_mem_req_02_p) {
+  nntrainer::MemoryPool pool;
+
+  pool.requestMemory(1, 4, 5);
+  EXPECT_EQ(pool.minMemoryRequirement(), 1);
```

Comment:
Could you checkout below patch before merge? In my gcc, there is sign compare error for these tests.

gcc (Ubuntu 9.3.0-11ubuntu0~18.04.1) 9.3.0

https://github.com/nnstreamer/nntrainer/pull/1543/commits/9c7104564adc1449a78b95cc7d647e43660b8eb1

Reply:
Is this the only place? I will update.

Reply:
as such are along this file :)

Reply:
please confer below (tested at ef1a259d)

<details>

```
[187/319] Compiling C++ object test/unittest/memory/libnntrainer_memory_planner_validation.a.p/memory_planner_validate.cpp.o
In file included from ../test/unittest/memory/memory_planner_validate.cpp:17:
/usr/src/gtest/include/gtest/gtest.h: In instantiation of ‘testing::AssertionResult testing::internal::CmpHelperEQ(const char*, const char*, const T1&, const T2&) [with T1 = long unsigned int; T2 = int]’:
/usr/src/gtest/include/gtest/gtest.h:1421:23:   required from ‘static testing::AssertionResult testing::internal::EqHelper<lhs_is_null_literal>::Compare(const char*, const char*, const T1&, const T2&) [with T1 = long unsigned int; T2 = int; bool lhs_is_null_literal = false]’
../test/unittest/memory/memory_planner_validate.cpp:54:3:   required from here
/usr/src/gtest/include/gtest/gtest.h:1392:11: warning: comparison of integer expressions of different signedness: ‘const long unsigned in’ and ‘const int’ [-Wsign-compare]
 1392 |   if (lhs == rhs) {
      |       ~~~~^~~~~~
[202/319] Compiling C++ object test/unittest/memory/unittest_memory.p/unittest_memory_pool.cpp.o
In file included from ../test/unittest/memory/unittest_memory_pool.cpp:18:
/usr/src/gtest/include/gtest/gtest.h: In instantiation of ‘testing::AssertionResult testing::internal::CmpHelperEQ(const char*, const char*, const T1&, const T2&) [with T1 = int; T2 = long unsigned int]’:
/usr/src/gtest/include/gtest/gtest.h:1421:23:   required from ‘static testing::AssertionResult testing::internal::EqHelper<lhs_is_null_literal>::Compare(const char*, const char*, const T1&, const T2&) [with T1 = int; T2 = long unsigned int; bool lhs_is_null_literal = false]’
../test/unittest/memory/unittest_memory_pool.cpp:96:3:   required from here
/usr/src/gtest/include/gtest/gtest.h:1392:11: warning: comparison of integer expressions of different signedness: ‘const int’ and ‘const long unsigned int’ [-Wsign-compare]
 1392 |   if (lhs == rhs) {
      |       ~~~~^~~~~~
/usr/src/gtest/include/gtest/gtest.h: In instantiation of ‘testing::AssertionResult testing::internal::CmpHelperEQ(const char*, const char*, const T1&, const T2&) [with T1 = long unsigned int; T2 = int]’:
/usr/src/gtest/include/gtest/gtest.h:1421:23:   required from ‘static testing::AssertionResult testing::internal::EqHelper<lhs_is_null_literal>::Compare(const char*, const char*, const T1&, const T2&) [with T1 = long unsigned int; T2 = int; bool lhs_is_null_literal = false]’
../test/unittest/memory/unittest_memory_pool.cpp:186:3:   required from here
/usr/src/gtest/include/gtest/gtest.h:1392:11: warning: comparison of integer expressions of different signedness: ‘const long unsigned in’ and ‘const int’ [-Wsign-compare]
```

</details>

Reply:
done.